Crash when running on lower versions of Android

Seems some classes are missing when using the current compile version. Please see details below.
https://stackoverflow.com/questions/45232457/android-studio-notification-channel-cannot-be-imported

```
2021-09-24 11:41:47.410 10073-10073/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.luxoft.sdl_core, PID: 10073
    java.lang.NoClassDefFoundError: Failed resolution of: Landroid/app/NotificationChannel;
        at org.luxoft.sdl_core.SdlLauncherService.enterForeground(SdlLauncherService.java:167)
        at org.luxoft.sdl_core.SdlLauncherService.onStartCommand(SdlLauncherService.java:55)
        at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:3522)
        at android.app.ActivityThread.-wrap23(ActivityThread.java)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1685)
        at android.os.Handler.dispatchMessage(Handler.java:105)
        at android.os.Looper.loop(Looper.java:156)
        at android.app.ActivityThread.main(ActivityThread.java:6623)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:942)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:832)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "android.app.NotificationChannel" on path: DexPathList[[zip file "/data/app/org.luxoft.sdl_core-1/base.apk"],nativeLibraryDirectories=[/data/app/org.luxoft.sdl_core-1/lib/arm64, /data/app/org.luxoft.sdl_core-1/base.apk!/lib/arm64-v8a, /system/lib64, /vendor/lib64, /system/vendor/lib64, /product/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:380)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        at org.luxoft.sdl_core.SdlLauncherService.enterForeground(SdlLauncherService.java:167) 
        at org.luxoft.sdl_core.SdlLauncherService.onStartCommand(SdlLauncherService.java:55) 
        at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:3522) 
        at android.app.ActivityThread.-wrap23(ActivityThread.java) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1685) 
        at android.os.Handler.dispatchMessage(Handler.java:105) 
        at android.os.Looper.loop(Looper.java:156) 
        at android.app.ActivityThread.main(ActivityThread.java:6623) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:942) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:832) 
```

--- a/app/src/main/java/org/luxoft/sdl_core/AndroidTool.java
+++ b/app/src/main/java/org/luxoft/sdl_core/AndroidTool.java
@@ -1,0 +1,24 @@
+package org.luxoft.sdl_core;
+
+import android.content.Context;
+import android.content.Intent;
+import android.os.Build;
+
+import androidx.annotation.NonNull;
+
+public class AndroidTool {
+
+    /**
+     * Service needs to be foregrounded in Android O and above.
+     * This will prevent apps in the background from crashing when they try to start SdlService,
+     * because Android O doesn't allow background apps to start background services.
+     * */
+    public static void startService(@NonNull Context context, @NonNull Intent intent) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            context.startForegroundService(intent);
+        } else {
+            context.startService(intent);
+        }
+    }
+
+}

--- a/app/src/main/java/org/luxoft/sdl_core/AndroidTool.java
+++ b/app/src/main/java/org/luxoft/sdl_core/AndroidTool.java
@@ -12,6 +12,9 @@ public class AndroidTool {
      * Service needs to be foregrounded in Android O and above.
      * This will prevent apps in the background from crashing when they try to start SdlService,
      * because Android O doesn't allow background apps to start background services.
+     *
+     * @param context   The object to use to start service.
+     * @param intent    The object to identify the service to be started. The Intent must be fully explicit.
      * */
     public static void startService(@NonNull Context context, @NonNull Intent intent) {
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {

--- a/app/src/main/java/org/luxoft/sdl_core/MainActivity.java
+++ b/app/src/main/java/org/luxoft/sdl_core/MainActivity.java
@@ -4,6 +4,7 @@ import static org.luxoft.sdl_core.BleCentralService.ACTION_SCAN_BLE;
 import static org.luxoft.sdl_core.BleCentralService.ACTION_START_BLE;
 import static org.luxoft.sdl_core.BleCentralService.ACTION_STOP_BLE;
 import static org.luxoft.sdl_core.BleCentralService.ON_BLE_SCAN_STARTED;
+import static org.luxoft.sdl_core.SdlLauncherService.IS_SDL_SERVICE_THREAD_ALIVE;
 import static org.luxoft.sdl_core.SdlLauncherService.ON_SDL_SERVICE_STARTED;
 import static org.luxoft.sdl_core.SdlLauncherService.ON_SDL_SERVICE_STOPPED;
 
@@ -67,8 +68,10 @@ public class MainActivity extends AppCompatActivity {
         start_sdl_button = findViewById(R.id.start_sdl_button);
         stop_sdl_button = findViewById(R.id.stop_sdl_button);
 
-        start_sdl_button.setEnabled(true);
-        stop_sdl_button.setEnabled(false);
+        Intent intent = getIntent();
+        boolean isSdlThreadAlive = intent != null && intent.getBooleanExtra(IS_SDL_SERVICE_THREAD_ALIVE, false);
+        start_sdl_button.setEnabled(!isSdlThreadAlive);
+        stop_sdl_button.setEnabled(isSdlThreadAlive);
 
         start_sdl_button.setOnClickListener(new View.OnClickListener() {
             @Override
@@ -114,7 +117,7 @@ public class MainActivity extends AppCompatActivity {
             }
         }
 
-        if (savedInstanceState == null) {
+        if (savedInstanceState == null && !isSdlThreadAlive) {
             initBT();
         }
         if (isBleSupported() && isBluetoothPermissionGranted()) {

--- a/app/src/main/java/org/luxoft/sdl_core/MainActivity.java
+++ b/app/src/main/java/org/luxoft/sdl_core/MainActivity.java
@@ -4,7 +4,6 @@ import static org.luxoft.sdl_core.BleCentralService.ACTION_SCAN_BLE;
 import static org.luxoft.sdl_core.BleCentralService.ACTION_START_BLE;
 import static org.luxoft.sdl_core.BleCentralService.ACTION_STOP_BLE;
 import static org.luxoft.sdl_core.BleCentralService.ON_BLE_SCAN_STARTED;
-import static org.luxoft.sdl_core.SdlLauncherService.IS_SDL_SERVICE_THREAD_ALIVE;
 import static org.luxoft.sdl_core.SdlLauncherService.ON_SDL_SERVICE_STARTED;
 import static org.luxoft.sdl_core.SdlLauncherService.ON_SDL_SERVICE_STOPPED;
 
@@ -68,10 +67,8 @@ public class MainActivity extends AppCompatActivity {
         start_sdl_button = findViewById(R.id.start_sdl_button);
         stop_sdl_button = findViewById(R.id.stop_sdl_button);
 
-        Intent intent = getIntent();
-        boolean isSdlThreadAlive = intent != null && intent.getBooleanExtra(IS_SDL_SERVICE_THREAD_ALIVE, false);
-        start_sdl_button.setEnabled(!isSdlThreadAlive);
-        stop_sdl_button.setEnabled(isSdlThreadAlive);
+        start_sdl_button.setEnabled(true);
+        stop_sdl_button.setEnabled(false);
 
         start_sdl_button.setOnClickListener(new View.OnClickListener() {
             @Override
@@ -117,7 +114,7 @@ public class MainActivity extends AppCompatActivity {
             }
         }
 
-        if (savedInstanceState == null && !isSdlThreadAlive) {
+        if (savedInstanceState == null) {
             initBT();
         }
         if (isBleSupported() && isBluetoothPermissionGranted()) {

--- a/app/src/main/java/org/luxoft/sdl_core/MainActivity.java
+++ b/app/src/main/java/org/luxoft/sdl_core/MainActivity.java
@@ -77,7 +77,7 @@ public class MainActivity extends AppCompatActivity {
                 stop_sdl_button.setEnabled(false);
                 Intent start_intent = new Intent(MainActivity.this, SdlLauncherService.class);
                 start_intent.setAction(ACTION_SDL_SERVICE_START);
-                startService(start_intent);
+                AndroidTool.startService(MainActivity.this, start_intent);
 
                 if (isBleSupported() && isBluetoothPermissionGranted()) {
                     final Intent intent = new Intent(ACTION_START_BLE);
@@ -96,7 +96,7 @@ public class MainActivity extends AppCompatActivity {
               
                 Intent stop_intent = new Intent(MainActivity.this, SdlLauncherService.class);
                 stop_intent.setAction(ACTION_SDL_SERVICE_STOP);
-                startService(stop_intent);
+                AndroidTool.startService(MainActivity.this, stop_intent);
             }
         });
 

--- a/app/src/main/java/org/luxoft/sdl_core/SdlLauncherService.java
+++ b/app/src/main/java/org/luxoft/sdl_core/SdlLauncherService.java
@@ -1,6 +1,5 @@
 package org.luxoft.sdl_core;
 
-import android.annotation.SuppressLint;
 import android.app.Notification;
 import android.app.NotificationChannel;
 import android.app.NotificationManager;
@@ -22,6 +21,7 @@ public class SdlLauncherService extends Service {
 
     public final static String ACTION_SDL_SERVICE_START = "ACTION_SDL_SERVICE_START";
     public final static String ACTION_SDL_SERVICE_STOP = "ACTION_SDL_SERVICE_STOP";
+    public final static String IS_SDL_SERVICE_THREAD_ALIVE = "IS_SDL_SERVICE_THREAD_ALIVE";
     public final static String ON_SDL_SERVICE_STOPPED = "org.luxoft.sdl_core.ON_SDL_SERVICE_STOPPED";
     public final static String ON_SDL_SERVICE_STARTED = "org.luxoft.sdl_core.ON_SDL_SERVICE_STARTED";
 
@@ -149,8 +149,12 @@ public class SdlLauncherService extends Service {
 
     private Notification getServiceNotification(String text){
         // The PendingIntent to launch our activity if the user selects this notification
-        PendingIntent contentIntent = PendingIntent.getActivity(this,
-                0,new Intent(this, SdlLauncherService.class),0);
+        Intent intent = new Intent(this, MainActivity.class);
+        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
+        boolean isThreadAlive = sdl_thread_ != null && sdl_thread_.isAlive();
+        intent.putExtra(IS_SDL_SERVICE_THREAD_ALIVE, isThreadAlive);
+
+        PendingIntent contentIntent = PendingIntent.getActivity(this, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);
 
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
             if (channel_id == null) {

--- a/app/src/main/java/org/luxoft/sdl_core/SdlLauncherService.java
+++ b/app/src/main/java/org/luxoft/sdl_core/SdlLauncherService.java
@@ -21,7 +21,6 @@ public class SdlLauncherService extends Service {
 
     public final static String ACTION_SDL_SERVICE_START = "ACTION_SDL_SERVICE_START";
     public final static String ACTION_SDL_SERVICE_STOP = "ACTION_SDL_SERVICE_STOP";
-    public final static String IS_SDL_SERVICE_THREAD_ALIVE = "IS_SDL_SERVICE_THREAD_ALIVE";
     public final static String ON_SDL_SERVICE_STOPPED = "org.luxoft.sdl_core.ON_SDL_SERVICE_STOPPED";
     public final static String ON_SDL_SERVICE_STARTED = "org.luxoft.sdl_core.ON_SDL_SERVICE_STARTED";
 
@@ -151,10 +150,8 @@ public class SdlLauncherService extends Service {
         // The PendingIntent to launch our activity if the user selects this notification
         Intent intent = new Intent(this, MainActivity.class);
         intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
-        boolean isThreadAlive = sdl_thread_ != null && sdl_thread_.isAlive();
-        intent.putExtra(IS_SDL_SERVICE_THREAD_ALIVE, isThreadAlive);
 
-        PendingIntent contentIntent = PendingIntent.getActivity(this, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);
+        PendingIntent contentIntent = PendingIntent.getActivity(this, 0, intent, PendingIntent.FLAG_IMMUTABLE);
 
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
             if (channel_id == null) {

--- a/app/src/main/java/org/luxoft/sdl_core/SdlLauncherService.java
+++ b/app/src/main/java/org/luxoft/sdl_core/SdlLauncherService.java
@@ -62,19 +62,12 @@ public class SdlLauncherService extends Service {
 
         if (action.equals(ACTION_SDL_SERVICE_STOP)) {
             stopSdlThread();
+            exitForeground();
             stopSelfResult(startId);
             return START_NOT_STICKY;
         }
 
         return super.onStartCommand(intent, flags, startId);
-    }
-
-    @Override
-    public void onDestroy() {
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
-            stopForeground(true);
-        }
-        super.onDestroy();
     }
 
     private boolean startSdlThread() {
@@ -146,7 +139,7 @@ public class SdlLauncherService extends Service {
     private void updateServiceNotification(String text) {
         Notification notification = getServiceNotification(text);
         if (notification == null) {
-            Log.w(TAG, "Cannot update notification. Channel_id is: " + channel_id);
+            Log.w(TAG, "Cannot update notification");
             return;
         }
 
@@ -155,37 +148,65 @@ public class SdlLauncherService extends Service {
     }
 
     private Notification getServiceNotification(String text){
-        if (channel_id == null) {
-            return null;
-        }
-
         // The PendingIntent to launch our activity if the user selects this notification
         PendingIntent contentIntent = PendingIntent.getActivity(this,
                 0,new Intent(this, SdlLauncherService.class),0);
 
-        return new NotificationCompat.Builder(this, channel_id)
-                .setContentTitle(SDL_CONTENT_TITLE)
-                .setContentText(text)
-                .setOnlyAlertOnce(true) // so when data is updated don't make sound and alert in android 8.0+
-                .setOngoing(true)
-                .setSmallIcon(R.drawable.ic_launcher_foreground)
-                .setContentIntent(contentIntent)
-                .build();
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            if (channel_id == null) {
+                Log.w(TAG, "Cannot create notification. Create and save channel_id firstly");
+                return null;
+            }
+            return new NotificationCompat.Builder(this, channel_id)
+                    .setContentTitle(SDL_CONTENT_TITLE)
+                    .setContentText(text)
+                    .setOnlyAlertOnce(true) // so when data is updated don't make sound and alert in android 8.0+
+                    .setOngoing(true)
+                    .setSmallIcon(R.drawable.ic_launcher_foreground)
+                    .setContentIntent(contentIntent)
+                    .build();
+        } else {
+            return new NotificationCompat.Builder(this)
+                    .setContentTitle(SDL_CONTENT_TITLE)
+                    .setContentText(text)
+                    .setOngoing(true)
+                    .setSmallIcon(R.drawable.ic_launcher_foreground)
+                    .setContentIntent(contentIntent)
+                    .build();
+        }
     }
 
-    @SuppressLint("NewApi")
-    public void enterForeground() {
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
-            NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
-            if (notificationManager != null) {
-                if (channel_id == null) {
-                    NotificationChannel channel = new NotificationChannel(APP_ID, SERVICE_NAME, NotificationManager.IMPORTANCE_DEFAULT);
-                    notificationManager.createNotificationChannel(channel);
-                    channel_id = channel.getId();
-                }
+    private void enterForeground() {
+        NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
+        if (notificationManager == null) {
+            Log.w(TAG, "NotificationManager are not available. Notification skipped.");
+            return;
+        }
+        final String text = "SDL core is starting...";
 
-                startForeground(FOREGROUND_SERVICE_ID, getServiceNotification("SDL core is starting..."));
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            if (channel_id == null) {
+                NotificationChannel channel = new NotificationChannel(APP_ID, SERVICE_NAME, NotificationManager.IMPORTANCE_DEFAULT);
+                notificationManager.createNotificationChannel(channel);
+                channel_id = channel.getId();
             }
+
+            startForeground(FOREGROUND_SERVICE_ID, getServiceNotification(text));
+        } else {
+            updateServiceNotification(text);
+        }
+    }
+
+    private void exitForeground() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            stopForeground(true);
+        } else {
+            NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
+            if (notificationManager == null) {
+                Log.w(TAG, "NotificationManager are not available. Notification cannot be canceled.");
+                return;
+            }
+            notificationManager.cancel(FOREGROUND_SERVICE_ID);
         }
     }
 

--- a/app/src/main/java/org/luxoft/sdl_core/SdlLauncherService.java
+++ b/app/src/main/java/org/luxoft/sdl_core/SdlLauncherService.java
@@ -11,11 +11,14 @@ import android.content.Intent;
 import android.os.Build;
 import android.os.Handler;
 import android.os.IBinder;
+import android.util.Log;
 import android.widget.Toast;
 
 import androidx.core.app.NotificationCompat;
 
 public class SdlLauncherService extends Service {
+
+    public final static String TAG = SdlLauncherService.class.getSimpleName();
 
     public final static String ACTION_SDL_SERVICE_START = "ACTION_SDL_SERVICE_START";
     public final static String ACTION_SDL_SERVICE_STOP = "ACTION_SDL_SERVICE_STOP";
@@ -59,12 +62,19 @@ public class SdlLauncherService extends Service {
 
         if (action.equals(ACTION_SDL_SERVICE_STOP)) {
             stopSdlThread();
-            stopForeground(true);
             stopSelfResult(startId);
             return START_NOT_STICKY;
         }
 
         return super.onStartCommand(intent, flags, startId);
+    }
+
+    @Override
+    public void onDestroy() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            stopForeground(true);
+        }
+        super.onDestroy();
     }
 
     private boolean startSdlThread() {
@@ -135,6 +145,10 @@ public class SdlLauncherService extends Service {
 
     private void updateServiceNotification(String text) {
         Notification notification = getServiceNotification(text);
+        if (notification == null) {
+            Log.w(TAG, "Cannot update notification. Channel_id is: " + channel_id);
+            return;
+        }
 
         NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
         notificationManager.notify(FOREGROUND_SERVICE_ID, notification);
@@ -161,15 +175,17 @@ public class SdlLauncherService extends Service {
 
     @SuppressLint("NewApi")
     public void enterForeground() {
-        NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
-        if (notificationManager != null) {
-            if (channel_id == null) {
-                NotificationChannel channel = new NotificationChannel(APP_ID, SERVICE_NAME, NotificationManager.IMPORTANCE_DEFAULT);
-                notificationManager.createNotificationChannel(channel);
-                channel_id = channel.getId();
-            }
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
+            if (notificationManager != null) {
+                if (channel_id == null) {
+                    NotificationChannel channel = new NotificationChannel(APP_ID, SERVICE_NAME, NotificationManager.IMPORTANCE_DEFAULT);
+                    notificationManager.createNotificationChannel(channel);
+                    channel_id = channel.getId();
+                }
 
-            startForeground(FOREGROUND_SERVICE_ID, getServiceNotification("SDL core is starting..."));
+                startForeground(FOREGROUND_SERVICE_ID, getServiceNotification("SDL core is starting..."));
+            }
         }
     }
 

